### PR TITLE
feat: add filter support

### LINKED ISSUE #36: eth_api: Implement `new_filter` method
### **Description:**
We currently have a stub for the [`new_filter`](https://github.com/matter-labs/era-test-node/blob/3539ad0a87f5cc5350b346bace874f4d966714b3/src/node.rs#L989) method, which is expected to handle creating a new filter based on a given criteria. We need to fully implement this method, ensure it's tested properly, and documented.

```rust
/// Creates a filter object, based on filter options, to notify when the state changes (logs).
fn new_filter(&self, _filter: Filter) -> jsonrpc_core::BoxFuture<jsonrpc_core::Result<U256>> {
    not_implemented("new_filter")
}
```

#### Expected response
Command:
```
# Note: Using mainnet URL for demonstration purposes
curl -X POST https://mainnet.era.zksync.io \
     -H "Content-Type: application/json" \
     --data '{
       "jsonrpc": "2.0",
       "id": 1,
       "method": "eth_newFilter",
       "params": [{
          "fromBlock": "0x1",
          "toBlock": "latest",
          "address": "0x11f943b2c77b743AB90f4A0Ae7d5A4e7FCA3E102",
          "topics": ["0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef"]
       }]
     }'
```
Response:
```
{"jsonrpc":"2.0","result":"0x247bc98d33cd6cfd9b2311cc5e129569a5bd4399d34fd1d456a78eeedebb4cfa","id":1}
```

### **Requirements:**

1. **Implement the Method:**
    - The method should take a `Filter` as an argument.
    - The method should create a new filter based on the given criteria and return a `U256` id for the created filter.

2. **Unit Tests:**
    - Write unit tests ensuring the `new_filter` method is working correctly.
    - Ensure edge cases are tested.
    - Ensure failures and errors are handled appropriately.

3. **Inline Documentation:**
    - Add inline Rust documentation (`///` comments) to describe the purpose and usage of the `new_filter` method.
    - Include any relevant information about parameters, return values, and potential errors/exceptions.

### **Task Checklist:**
- [ ] Implement the `new_filter` method.
- [ ] Write unit tests for the `new_filter` method.
- [ ] Add inline documentation for the `new_filter` method.

### **Additional Notes:**
If you're an external contributor looking to take on this issue, please comment below to let us know you're working on it. If you need further clarification on the requirements or need assistance with anything related to the implementation, feel free to ask!

### LINKED ISSUE #37: eth_api: Implement `new_block_filter` method
### **Description:**
We currently have a stub for the [`new_block_filter`](https://github.com/matter-labs/era-test-node/blob/3539ad0a87f5cc5350b346bace874f4d966714b3/src/node.rs#L993) method, which is expected to handle the creation of a new filter for receiving notifications when a new block is mined. We need to fully implement this method, ensure it's tested properly, and documented.

```rust
/// Creates a filter that notifies when a new block is mined.
fn new_block_filter(&self) -> jsonrpc_core::BoxFuture<jsonrpc_core::Result<U256>> {
    not_implemented("new_block_filter")
}
```

#### Expected response
Command:
```curl
# Note: Using mainnet URL for demonstration purposes
curl -X POST https://mainnet.era.zksync.io \
     -H "Content-Type: application/json" \
     --data '{
       "jsonrpc": "2.0",
       "id": 1,
       "method": "eth_newBlockFilter",
       "params": []
     }'
```
Response:
```
{"jsonrpc":"2.0","result":"0x72df37711637fcd85355867f08abc64c17dc04719d84da846a3a9f9ca22c4265","id":1}
```

### **Requirements:**

1. **Implement the Method:**
    - The method should not take any arguments.
    - It should create a new filter that notifies the caller when a new block is mined.
    - Return a `U256` id for the created filter.

2. **Unit Tests:**
    - Write unit tests ensuring the `new_block_filter` method is working correctly.
    - Ensure edge cases are tested.
    - Ensure failures and errors are handled appropriately.

3. **Inline Documentation:**
    - Add inline Rust documentation (`///` comments) to describe the purpose and usage of the `new_block_filter` method.
    - Include any relevant information about the return values and potential errors/exceptions.

### **Task Checklist:**
- [ ] Implement the `new_block_filter` method.
- [ ] Write unit tests for the `new_block_filter` method.
- [ ] Add inline documentation for the `new_block_filter` method.

### **Additional Notes:**
If you're an external contributor looking to take on this issue, please comment below to let us know you're working on it. If you need further clarification on the requirements or need assistance with anything related to the implementation, feel free to ask!

### LINKED ISSUE #38: eth_api: Implement `uninstall_filter` method
### **Description:**
We currently have a placeholder for the [`uninstall_filter`](https://github.com/matter-labs/era-test-node/blob/3539ad0a87f5cc5350b346bace874f4d966714b3/src/node.rs#L997C8-L997C24) method, which is anticipated to handle the removal of an existing filter based on a provided filter ID. 

```rust
/// Uninstalls a previously created filter by its ID.
fn uninstall_filter(&self, filter_id: U256) -> jsonrpc_core::BoxFuture<jsonrpc_core::Result<bool>> {
    not_implemented("uninstall_filter")
}
```

#### Expected response

Command:
```
# Note: Using mainnet URL for demonstration purposes
curl -X POST https://mainnet.era.zksync.io \
     -H "Content-Type: application/json" \
     --data '{
       "jsonrpc": "2.0",
       "id": 1,
       "method": "eth_uninstallFilter",
       "params": ["0x72df37711637fcd85355867f08abc64c17dc04719d84da846a3a9f9ca22c4265"]
     }'
```
Response:
```
{"jsonrpc":"2.0","result":false,"id":1}
```

### **Requirements:**

1. **Implement the Method:**
    - The method should take a `U256` filter ID as its argument.
    - Remove the filter associated with the provided ID.
    - Return a boolean indicating the success or failure of the uninstallation.

2. **Unit Tests:**
    - Write unit tests ensuring the `uninstall_filter` method operates correctly.
    - Address potential edge cases in your testing.
    - Implement error-handling measures for potential failures.

3. **Inline Documentation:**
    - Include inline Rust documentation (`///` comments) to outline the purpose and application of the `uninstall_filter` method.
    - Provide details regarding parameters, return values, and any potential errors or exceptions.

### **Task Checklist:**
- [ ] Implement the `uninstall_filter` method.
- [ ] Write unit tests for the `uninstall_filter` method.
- [ ] Include inline documentation for the `uninstall_filter` method.

### **Additional Notes:**
If you're an external contributor looking to take on this issue, please comment below to let us know you're working on it. If you need further clarification on the requirements or need assistance with anything related to the implementation, feel free to ask!

### LINKED ISSUE #39: eth_api: Implement `new_pending_transaction_filter` method
### **Description:**
We currently have a placeholder for the [`new_pending_transaction_filter`](https://github.com/matter-labs/era-test-node/blob/3539ad0a87f5cc5350b346bace874f4d966714b3/src/node.rs#L1001C8-L1001C38) method, which is designed to establish a filter that notifies when new pending transactions enter the transaction pool. 

```rust
/// Creates a filter to notify when new pending transactions arise.
fn new_pending_transaction_filter(&self) -> jsonrpc_core::BoxFuture<jsonrpc_core::Result<U256>> {
    not_implemented("new_pending_transaction_filter")
}
```

#### Expected response

Command:
```
# Note: Using mainnet URL for demonstration purposes
curl -X POST https://mainnet.era.zksync.io \
     -H "Content-Type: application/json" \
     --data '{
       "jsonrpc": "2.0",
       "id": 1,
       "method": "eth_newPendingTransactionFilter",
       "params": []
     }'
```
Response:
```
{"jsonrpc":"2.0","result":"0xf86f90e83fd8a20ed75b6f7ed364c301a551abc98d5318d8d12e5ea4de7aa15a","id":1}
```

### **Requirements:**

1. **Implement the Method:**
    - The method doesn't require any arguments.
    - It should establish a new filter and return a `U256` id for the created filter that's oriented towards pending transactions.

2. **Unit Tests:**
    - Write unit tests to confirm the `new_pending_transaction_filter` method's functionality.
    - Consider and test for potential edge cases.
    - Ensure that potential errors and exceptions are managed appropriately.

3. **Inline Documentation:**
    - Add inline Rust documentation (`///` comments) to define the purpose and implementation of the `new_pending_transaction_filter` method.
    - Include any relevant information about the return values and potential errors/exceptions.

### **Task Checklist:**
- [ ] Implement the `new_pending_transaction_filter` method.
- [ ] Write unit tests for the `new_pending_transaction_filter` method.
- [ ] Include inline documentation for the `new_pending_transaction_filter` method.

### **Additional Notes:**
If you're an external contributor looking to take on this issue, please comment below to let us know you're working on it. If you need further clarification on the requirements or need assistance with anything related to the implementation, feel free to ask!

### LINKED ISSUE #42: eth_api: Implement `get_filter_changes` method
### **Description:**
We need to implement the [`get_filter_changes`](https://github.com/matter-labs/era-test-node/blob/3539ad0a87f5cc5350b346bace874f4d966714b3/src/node.rs#L1021C8-L1021C26) method. This method aims to retrieve the changes associated with a specific filter ID, ensuring a more dynamic interaction. Polling method for a filter, which returns an array of logs which occurred since last poll. Filter must be created by calling either [eth_newFilter](https://docs.infura.io/networks/ethereum/json-rpc-methods/filter-methods/eth_newfilter) or [eth_newBlockFilter](https://docs.infura.io/networks/ethereum/json-rpc-methods/filter-methods/eth_newblockfilter).

```rust
/// Retrieves the changes associated with a specific filter ID.
fn get_filter_changes(&self, filter_id: U256) -> jsonrpc_core::BoxFuture<jsonrpc_core::Result<Vec<Change>>> {
    not_implemented("get_filter_changes")
}
```

### **Requirements:**

1. **Implement the Method:**
    - This method should take in a `U256` filter ID.
    - Based on this ID, it should return the changes in a `Vec<Change>` format associated with the filter.

2. **Unit Tests:**
    - Write unit tests to ensure the correct operation of the `get_filter_changes` method.
    - Ensure that you cover various edge cases.
    - Ensure failures and exceptions are handled aptly.

3. **Inline Documentation:**
    - Include inline Rust documentation (`///` comments) detailing the function and utility of the `get_filter_changes` method.
    - Include any relevant information about the return values and potential errors/exceptions.

### **Task Checklist:**
- [ ] Implement the `get_filter_changes` method.
- [ ] Write unit tests for the `get_filter_changes` method.
- [ ] Include inline documentation for the `get_filter_changes` method.

### **Additional Notes:**
If you are an external contributor and are considering working on this issue, please comment below so we are aware. Should you need clarification on any aspect of the requirements or seek assistance regarding the implementation, do not hesitate to ask.

### PR DESCRIPTION
**Pull request by nbaztec**
# What :computer: 
* Adds support for `eth_newBlockFilter`, `eth_newFilter`, `eth_newPendingTransactionFilter`, `eth_uninstallFilter`, and `eth_getFilterChanges`.
* Keeps track of transaction logs in node.

# Why :hand:
* Filters are a quite common use case that enables developers to get the exact information they are interested in, quickly.

# Evidence :camera:
image
output

# Notes :memo:
* Fixes #36 
* Fixes #37
* Fixes #38
* Fixes #39
* Fixes #42

### CODE REVIEW

File: src/filters.rs [hunk 1]
```diff
@@ -0,0 +1,1115 @@
...
+use zksync_web3_decl::types::FilterChanges;
+
+/// Specifies a filter type
+#[derive(Debug, Clone, PartialEq)]
+pub enum FilterType {
+    /// A filter for block information
+    Block(BlockFilter),
+    /// A filter for log information.
+    /// This is [Box] to ensure the enum invariants are similar size.
+    Log(Box<LogFilter>),
```
Comment (nbaztec) [inline src/filters.rs:14]: I'm a bit on the fence with this warning. I get the utility of it and in general agree to it, but this doesn't seem to be that future safe if for some reason _all_ invariants become huge. Admittedly in our case we know this won't happen, but I just found it a bit weird :D 

image

File: src/filters.rs [hunk 2]
```diff
@@ -0,0 +1,1115 @@
...
+    from_block: BlockNumber,
+    to_block: BlockNumber,
+    addresses: Vec<H160>,
+    topics: [Option<HashSet<H256>>; 4],
+    updates: Vec<Log>,
+}
+
+impl LogFilter {
+    fn matches(&self, log: &Log, latest_block_number: U64) -> bool {
+        let from = match self.from_block {
```
Comment (MexicanAce) [inline src/filters.rs:37]: Should we create a helper function in `utils.rs` that returns `u64` given `BlockNumber`? I think we've used this in a few places now.

File: src/node.rs [hunk 3]
```diff
@@ -1133,18 +1152,87 @@ impl<S: ForkSource + std::fmt::Debug> InMemoryNode<S> {
...
+                info: TxExecutionInfo {
+                    tx: l2_tx,
+                    batch_number: block.l1_batch_number.unwrap_or_default().as_u32(),
+                    miniblock_number: current_miniblock,
+                    result,
+                },
+                receipt: tx_receipt,
             },
         );
+        let block_hash = block.hash;
```
Comment (MexicanAce) [inline src/node.rs:1231]: Is this variable necessary?

File: src/node.rs [hunk 4]
```diff
@@ -1109,6 +1119,15 @@ impl<S: ForkSource + std::fmt::Debug> InMemoryNode<S> {
         let tx_hash = l2_tx.hash();
         log::info!("");
         log::info!("Executing {}", format!("{:?}", tx_hash).bold());
+
+        {
+            let mut inner = self
+                .inner
+                .write()
+                .map_err(|e| format!("Failed to acquire write lock: {}", e))?;
+            inner.filters.notify_new_pending_transaction(l2_tx.hash());
```
Comment (MexicanAce) [inline src/node.rs:1128]: nit: This can be simplified to use the `tx_hash` variable
```suggestion
            inner.filters.notify_new_pending_transaction(tx_hash);
```

File: src/filters.rs [hunk 5]
```diff
@@ -0,0 +1,1115 @@
...
+                    None,
+                    Some(hashset! {}),
+                ],
+            )
+            .expect("failed adding filter");
+
+        assert_eq!(U256::from(1), id);
+        assert_eq!(
+            hashmap! {
+                U256::from(1) => FilterType::Log(LogFilter {
```
Comment (MexicanAce) [inline src/filters.rs:272]: For some reason this is giving me an inline error in VSCode, but it compiles fine... any ideas?
image

Comment (nbaztec) [reply src/filters.rs:272]: That's not you, I missed boxing this in the test for the same reason it compiles - it is a test.

File: src/node.rs (hunk 3 above)
Comment (nbaztec) [reply src/node.rs:1231]: The variable `block` is inserted into the `blocks` storage so the hash isn't available when we need to notify it on line `1234`. The alternate would be to clone the entire variable.

File: src/filters.rs (hunk 1 above)
Comment (MexicanAce) [reply src/filters.rs:14]: I'm still a newbie with Rust, so I just went down a rabbit hole of variant sizes and `Box<>`... at this point I'm deferring the almighty clippy, but agreed that this is weird

File: src/filters.rs (hunk 5 above)
Comment (MexicanAce) [reply src/filters.rs:272]: TIL that Rust ignores test code in `.rs` files when compiling

File: src/filters.rs (hunk 1 above)
Comment (nbaztec) [reply src/filters.rs:14]: It makes total sense what clippy suggests, since the size of an enum/union is as big as the biggest invariant. In our case each `Log` object would be `272 bytes` even if we store 100x `Block` invariant which is a mere `24 bytes`. This can cause a very high memory usage for no good reason when storing a significant number of objects. Therefore, clippy suggests us to allocate `LogFilter` on the heap where it will only occupy `usize` worth of size for the pointer - mitigating the issue.

But..... I know for a fact this will never be significant for us, and is almost certainly a premature optimization at best and a code smell at worst (since this isn't intuitive). Hence, my apprehension, but then again I'm also not a huge fan of adding ignores to linter thus I went with the comment explanation.

File: src/filters.rs (hunk 5 above)
Comment (nbaztec) [reply src/filters.rs:272]: yeah, it needs the `--tests` with `cargo check` and `cargo build` to force it :)

File: src/filters.rs (hunk 2 above)
Comment (nbaztec) [reply src/filters.rs:37]: I don't recall me adding this, can you perhaps give an example? In my case I needed to pass the `latest_block_number` since there's no way to map `latest` without this knowledge.

Comment (MexicanAce) [reply src/filters.rs:37]: Hmm, thought I needed to do this with gas estimation at some point, but looks like the only other places are for `get_transaction_count_by_number` and `get_block_by_number`.

Though it would be a slight refactor/optimization for `get_block_by_number`.

And since it would be a helper function, I think it would make sense to pass in `latest_block_number` instead of a `reader`.

Comment (MexicanAce) [reply src/filters.rs:37]: Either way, not necessary for this PR (resolving conversation)

Comment (nbaztec) [reply src/filters.rs:37]: already committed it
